### PR TITLE
Add a Params object to carry parsed and validated Config

### PR DESCRIPTION
**Pull request by BrettMichaelJohnson**

### CODE REVIEW

File: src/com/google/enterprise/adaptor/filenet/FileNetAdaptor.java [hunk 1]
```diff
@@ -83,46 +82,44 @@ public void initConfig(Config config) {
     config.addKey("filenet.username", null);
     config.addKey("filenet.password", null);
     config.addKey("filenet.objectStore", null);
+    config.addKey("filenet.objectFactory", null);
```
Comment (wiarlawd) [inline src/com/google/enterprise/adaptor/filenet/FileNetAdaptor.java]: The ObjectFactory class name is internal, and should default to FileNetObjectFactory.

File: src/com/google/enterprise/adaptor/filenet/FileNetAdaptor.java [hunk 2]
```diff
@@ -62,10 +63,8 @@
   private final ObjectFactory factory;
 
   private AdaptorContext context;
-  private String contentEngineUrl;
-  private String username;
-  private String password;
-  private String objectStore;
+  private Params params;
```
Comment (wiarlawd) [inline src/com/google/enterprise/adaptor/filenet/FileNetAdaptor.java]: Rename to something: ConfigOptions, Options, ConfigWrapper, Parameters.

File: src/com/google/enterprise/adaptor/filenet/FileNetAdaptor.java [hunk 3]
```diff
@@ -83,46 +82,44 @@ public void initConfig(Config config) {
...
-    password = config.getValue("filenet.password");
-
-    // Verify we can connect to the server and access the ObjectStore.
-    logger.log(Level.INFO, "Connecting to content engine {0}",
-        contentEngineUrl);
     try (Connection connection = getConnection()) {
-      logger.log(Level.INFO, "Connecting to object store {0}", objectStore);
-      factory.getObjectStore(connection, objectStore);
+      logger.log(Level.INFO, "Connecting to object store {0}",
+          params.getObjectStore());
```
Comment (wiarlawd) [inline src/com/google/enterprise/adaptor/filenet/FileNetAdaptor.java]: I think the Params should get the factory class name, and hold both the actual ObjectFactory and possibly have getters for some of the created objects. FileSession in v3 had a getter for the ObjectFactory and created and passed the ObjectStore to the traversers. I think we want getConnection from this clas to move to Params as well. Some fuzziness abounds about the wiring, but getting the ObjectFactory and the Connection from the Params, and getting the rest either from the Params or the ObjectFactory (maybe depending on whether constructing them requires values from the Params?).

File: src/com/google/enterprise/adaptor/filenet/Params.java [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+    contentEngineUrl = config.getValue("filenet.contentEngineUrl");
+    logger.log(Level.CONFIG, "filenet.contentEngineUrl: {0}", contentEngineUrl);
+    try {
+      new ValidatedUri(contentEngineUrl).logUnreachableHost();
+    } catch (URISyntaxException e) {
+      throw new InvalidConfigurationException(
+          "Invalid filenet.contentEngineUrl: " + e.getMessage());
+    }
+
+    objectStore = config.getValue("filenet.objectStore").trim();
```
Comment (wiarlawd) [inline src/com/google/enterprise/adaptor/filenet/Params.java]: Trim is redundant, the library trims all config values since 4.0.4 or something like that.

File: src/com/google/enterprise/adaptor/filenet/Params.java [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+    if (objectStore.isEmpty()) {
+      throw new InvalidConfigurationException("filenet.objectStore may not be empty");
+    }
+    logger.log(Level.CONFIG, "filenet.objectStore: {0}", objectStore);
+
+    username = config.getValue("filenet.username");
+    password = config.getValue("filenet.password");
+
+    // TODO(bmj): load the object factor?
```
Comment (wiarlawd) [inline src/com/google/enterprise/adaptor/filenet/Params.java]: Yes, please.

File: src/com/google/enterprise/adaptor/filenet/Params.java [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+    objectFactory = config.getValue("filenet.objectFactory").trim();
+    if (objectFactory.isEmpty()) {
+      throw new InvalidConfigurationException(
+          "filenet.objectFactory may not be empty");
+    }
+    logger.log(Level.CONFIG, "filenet.objectFactory: {0}", objectFactory);
+
+    // TODO(bmj): If empty, can I base it off of contentEngineUrl?
```
Comment (wiarlawd) [inline src/com/google/enterprise/adaptor/filenet/Params.java]: No.

File: src/com/google/enterprise/adaptor/filenet/Params.java [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+    displayUrl = workplaceUrl;
+    logger.log(Level.CONFIG, "displayUrl: {0}", displayUrl);
+    try {
+      new ValidatedUri(displayUrl + "0").logUnreachableHost();
+    } catch (URISyntaxException e) {
+      throw new InvalidConfigurationException(
+          "Invalid displayUrl: " + e.getMessage());
+    }
+
+    isPublic = Boolean.parseBoolean(config.getValue("filenet.isPublic").trim());
```
Comment (wiarlawd) [inline src/com/google/enterprise/adaptor/filenet/Params.java]: I think this gets replaced by adaptor.markAllDocsAsPublic.

File: src/com/google/enterprise/adaptor/filenet/Params.java [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+    logger.log(Level.CONFIG, "filenet.isPublic: " + isPublic);
+
+    String pushAclsStr = config.getValue("filenet.pushAcls").trim();
```
Comment (wiarlawd) [inline src/com/google/enterprise/adaptor/filenet/Params.java]: I think this goes away, too. Do any of the v4 connectors making pushing ACLs optional?

File: src/com/google/enterprise/adaptor/filenet/FileNetAdaptor.java [hunk 9]
```diff
@@ -28,6 +28,7 @@
 import com.google.enterprise.adaptor.DocId;
 import com.google.enterprise.adaptor.DocIdPusher;
 import com.google.enterprise.adaptor.InvalidConfigurationException;
+import com.google.enterprise.adaptor.Principal;
```
Comment (wiarlawd) [inline src/com/google/enterprise/adaptor/filenet/FileNetAdaptor.java]: I would definitely statically import DEFAULT_NAMESPACE rather than not import the java.security.Principal class when both are used. If we need more from Principal then we can deal with it, but we probably don't (no v4 connector currently does, AFAIK).

File: src/com/google/enterprise/adaptor/filenet/FileNetAdaptor.java (hunk 3 above)
Comment (BrettMichaelJohnson) [inline src/com/google/enterprise/adaptor/filenet/FileNetAdaptor.java]: Done.

File: src/com/google/enterprise/adaptor/filenet/ConfigOptions.java [hunk 10]
```diff
@@ -58,34 +65,46 @@ public Params(Config config) throws InvalidConfigurationException {
...
+    String objectFactoryName = config.getValue("filenet.objectFactory");
+    if (objectFactoryName.isEmpty()) {
       throw new InvalidConfigurationException(
           "filenet.objectFactory may not be empty");
     }
-    logger.log(Level.CONFIG, "filenet.objectFactory: {0}", objectFactory);
+    try {
+      objectFactory =
+          (ObjectFactory) Class.forName(objectFactoryName).newInstance();
+    } catch (InstantiationException e) {
```
Comment (wiarlawd) [inline src/com/google/enterprise/adaptor/filenet/ConfigOptions.java]: I know this is a copy from FileSession, but the verbosity for errors that can't happen is too high. Let's just | these together.

File: src/com/google/enterprise/adaptor/filenet/ConfigOptions.java [hunk 11]
```diff
@@ -97,14 +116,12 @@ public Params(Config config) throws InvalidConfigurationException {
           "Invalid displayUrl: " + e.getMessage());
     }
 
-    isPublic = Boolean.parseBoolean(config.getValue("filenet.isPublic").trim());
-    logger.log(Level.CONFIG, "filenet.isPublic: " + isPublic);
+    markAllDocsAsPublic =
+        Boolean.parseBoolean(config.getValue("adaptor.markAllDocsAsPublic"));
+    logger.log(Level.CONFIG, "adaptor.markAllDocsAsPublic"
```
Comment (wiarlawd) [inline src/com/google/enterprise/adaptor/filenet/ConfigOptions.java]: MessageFormat-style (plus missing space): s/" +/: {0}",/

File: src/com/google/enterprise/adaptor/filenet/ConfigOptions.java [hunk 12]
```diff
@@ -97,14 +116,12 @@ public Params(Config config) throws InvalidConfigurationException {
...
+        + markAllDocsAsPublic);
 
-    String pushAclsStr = config.getValue("filenet.pushAcls").trim();
-    pushAcls = pushAclsStr.isEmpty() ? true : Boolean.parseBoolean(pushAclsStr);
-    logger.log(Level.CONFIG, "filenet.pushAcls: " + pushAcls);
-
-    globalNamespace = config.getValue("adaptor.namespace").trim();
+    globalNamespace = config.getValue("adaptor.namespace");
     logger.log(Level.CONFIG, "adaptor.namespace: " + globalNamespace);
```
Comment (wiarlawd) [inline src/com/google/enterprise/adaptor/filenet/ConfigOptions.java]: Add missing {0} here and below.

File: src/com/google/enterprise/adaptor/filenet/FileNetAdaptor.java [hunk 13]
```diff
@@ -82,44 +82,33 @@ public void initConfig(Config config) {
     config.addKey("filenet.username", null);
     config.addKey("filenet.password", null);
     config.addKey("filenet.objectStore", null);
-    config.addKey("filenet.objectFactory", null);
+    config.addKey("filenet.objectFactory", FileNetObjectFactory.class.getName());
```
Comment (wiarlawd) [inline src/com/google/enterprise/adaptor/filenet/FileNetAdaptor.java]: 80 cols

File: src/com/google/enterprise/adaptor/filenet/ConfigOptions.java [hunk 14]
```diff
@@ -16,39 +16,46 @@
...
+  private final boolean markAllDocsAsPublic;
   private final String additionalWhereClause;
   private final String deleteAdditionalWhereClause;
   private final Set<String> includedMetadata;
   private final Set<String> excludedMetadata;
   private final String globalNamespace;
   private final int maxFeedUrls;
 
-  public Params(Config config) throws InvalidConfigurationException {
+  public ConfigOptions(AdaptorContext context) throws InvalidConfigurationException {
```
Comment (wiarlawd) [inline src/com/google/enterprise/adaptor/filenet/ConfigOptions.java]: 80 cols. Why the spate of 80 cols?

File: src/com/google/enterprise/adaptor/filenet/FileNetAdaptor.java [hunk 15]
```diff
@@ -82,44 +82,33 @@ public void initConfig(Config config) {
...
 
-    documentTraverser = new MockTraverser(params);
-  }
-
-  Params getParams() {
-    return params;
+    documentTraverser = new MockTraverser(configOptions);
   }
 
   @VisibleForTesting
```
Comment (wiarlawd) [inline src/com/google/enterprise/adaptor/filenet/FileNetAdaptor.java]: Exists for testing?

File: test/com/google/enterprise/adaptor/filenet/FileNetAdaptorTest.java [hunk 16]
```diff
@@ -99,21 +107,22 @@ private static DocId newDocId(String id) {
     return new DocId("guid/" + id);
   }
 
+  /* TODO(bmj): this test is broken wrt UserContext assertions
```
Comment (wiarlawd) [inline test/com/google/enterprise/adaptor/filenet/FileNetAdaptorTest.java]: Broken how? Can we use `@Ignore` instead of commenting it out?

File: test/com/google/enterprise/adaptor/filenet/FileNetAdaptorTest.java [hunk 17]
```diff
@@ -254,7 +263,9 @@ public void testInit_objectStore() throws Exception {
     String objectStore = "ObjectStore";
     config.overrideKey("filenet.objectStore", objectStore);
     adaptor.init(context);
-    assertEquals(objectStore, adaptor.getParams().getObjectStore());
+    Connection conn = getConfigOptions().getConnection();
+    // TODO(bmj): use hamcrest.isA
+    assertTrue(getConfigOptions().getObjectStore(conn) instanceof ObjectStore);
```
Comment (wiarlawd) [inline test/com/google/enterprise/adaptor/filenet/FileNetAdaptorTest.java]: This could hardly fail, since getObjectStore returns an ObjectStore. What is this supposed to be testing.

File: test/com/google/enterprise/adaptor/filenet/FileNetAdaptorTest.java [hunk 18]
```diff
@@ -345,113 +366,93 @@ public void testInit_displayUrl_missing() throws Exception {
...
-    assertEquals("namespace", adaptor.getParams().getGlobalNamespace());
+    assertEquals("namespace", getConfigOptions().getGlobalNamespace());
   }
 
   @Test
   public void testInit_globalNamespace_default() throws Exception {
     adaptor.init(context);
-    assertEquals(Principal.DEFAULT_NAMESPACE,
-        adaptor.getParams().getGlobalNamespace());
+    assertEquals(DEFAULT_NAMESPACE,
```
Comment (wiarlawd) [inline test/com/google/enterprise/adaptor/filenet/FileNetAdaptorTest.java]: Join lines?

File: src/com/google/enterprise/adaptor/filenet/ConfigOptions.java (hunk 14 above)
Comment (BrettMichaelJohnson) [inline src/com/google/enterprise/adaptor/filenet/ConfigOptions.java]: Global search and replace.

File: src/com/google/enterprise/adaptor/filenet/ConfigOptions.java (hunk 10 above)
Comment (BrettMichaelJohnson) [inline src/com/google/enterprise/adaptor/filenet/ConfigOptions.java]: Done.

File: src/com/google/enterprise/adaptor/filenet/ConfigOptions.java (hunk 11 above)
Comment (BrettMichaelJohnson) [inline src/com/google/enterprise/adaptor/filenet/ConfigOptions.java]: Done.

File: src/com/google/enterprise/adaptor/filenet/ConfigOptions.java (hunk 12 above)
Comment (BrettMichaelJohnson) [inline src/com/google/enterprise/adaptor/filenet/ConfigOptions.java]: Done.

File: src/com/google/enterprise/adaptor/filenet/FileNetAdaptor.java (hunk 13 above)
Comment (BrettMichaelJohnson) [inline src/com/google/enterprise/adaptor/filenet/FileNetAdaptor.java]: Done.

File: src/com/google/enterprise/adaptor/filenet/FileNetAdaptor.java (hunk 15 above)
Comment (BrettMichaelJohnson) [inline src/com/google/enterprise/adaptor/filenet/FileNetAdaptor.java]: Yes.

File: test/com/google/enterprise/adaptor/filenet/FileNetAdaptorTest.java (hunk 17 above)
Comment (BrettMichaelJohnson) [inline test/com/google/enterprise/adaptor/filenet/FileNetAdaptorTest.java]: Done.

File: test/com/google/enterprise/adaptor/filenet/FileNetAdaptorTest.java (hunk 18 above)
Comment (BrettMichaelJohnson) [inline test/com/google/enterprise/adaptor/filenet/FileNetAdaptorTest.java]: Done.

File: test/com/google/enterprise/adaptor/filenet/FileNetAdaptorTest.java [hunk 19]
```diff
@@ -97,15 +109,16 @@ private static DocId newDocId(String id) {
     return new DocId("guid/" + id);
   }
 
+  @Ignore // TODO(bmj): this test is broken wrt UserContext assertions
   @Test
   public void testGetConnection() throws Exception {
     // Mock sensitiveValueDecoder uppercases the value.
     Subject subject = new Subject(true, ImmutableSet.<Principal>of(),
-        ImmutableSet.of("test"), ImmutableSet.of("PASSWORD"));
+        ImmutableSet.of("bugger"), ImmutableSet.of("PASSWORD"));
```
Comment (wiarlawd) [inline test/com/google/enterprise/adaptor/filenet/FileNetAdaptorTest.java]: Maybe undo this change?

File: test/com/google/enterprise/adaptor/filenet/FileNetAdaptorTest.java [hunk 20]
```diff
@@ -97,15 +109,16 @@ private static DocId newDocId(String id) {
     return new DocId("guid/" + id);
   }
 
+  @Ignore // TODO(bmj): this test is broken wrt UserContext assertions
```
Comment (wiarlawd) [inline test/com/google/enterprise/adaptor/filenet/FileNetAdaptorTest.java]: `@Ignore` takes a string argument that is printed by "ant test", so you can just include the comment (TODO and all) in the annotation: `@Ignore("TODO(bmj): ...")`
